sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/executeImmediate.scala

```diff
@@ -0,0 +1,186 @@
+    either match {
+      case Left(v) => v
+      case Right(u) =>
+        val varReference = getVariableReference(u.nameParts)
+
+        if (!varReference.dataType.sameType(StringType)) {
+          throw new AnalysisException(
+            errorClass = "INVALID_VARIABLE_TYPE_FOR_QUERY_EXECUTE_IMMEDIATE",
+            messageParameters = Map("varType" -> varReference.dataType.simpleString))
+        }
+
+        // Call eval with null value passed instead of a row.
+        // This is ok as this is variable and invoking eval should
+        // be independent of row value.
+        varReference.eval(null).toString
+    }
+  }
+
+  override def apply(plan: LogicalPlan): LogicalPlan =
+    plan.resolveOperatorsWithPruning(_.containsPattern(EXECUTE_IMMEDIATE), ruleId) {
+      case ExecuteImmediateQuery(expressions, query, targetVariablesOpt, parser) =>
+        val queryString = extractQueryString(query)
+        val plan = parseStatement(parser, queryString, targetVariablesOpt)
+
+        val posNodes = plan.collect { case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect { case n: PosParameter => n })
+        }.flatten
+        val namedNodes = plan.collect { case p: LogicalPlan =>
+          p.expressions.flatMap(_.collect { case n: NamedParameter => n })
+        }.flatten
+
+        val queryPlan = if (expressions.isEmpty || (posNodes.isEmpty && namedNodes.isEmpty)) {
+          plan
+        } else if (posNodes.nonEmpty && namedNodes.nonEmpty) {
+          throw new AnalysisException(
+            errorClass = "INVALID_QUERY_BOTH_POSITIONAL_AND_NAMED_PARAMETERS_PRESENT",
+            messageParameters = Map.empty)
+        } else {
+          if (posNodes.nonEmpty) {
+            PosParameterizedQuery(
+              plan,
+              // We need to resolve arguments before Resolution batch to make sure
+              // that some rule does not accidently resolve our parameters.
+              // We do not want this as they can resolve some unsupported parameters
+              resolveArguments(expressions))
+          } else {
+            val namedExpressions = expressions.collect { case (e: NamedExpression) => e }
+
+            NameParameterizedQuery(
+              plan,
+              namedExpressions.map(_.name),
+              // We need to resolve arguments before Resolution batch to make sure
+              // that some rule does not accidently resolve our parameters.
+              // We do not want this as they can resolve some unsupported parameters.
+              resolveArguments(namedExpressions))
+          }
+        }
+
+        targetVariablesOpt
+          .map(variables => {
+            SetVariable(variables, queryPlan)
+          })
+          .getOrElse { queryPlan }
+    }
+
+  private def parseStatement(
+    parser: ParserInterface,
+    queryString: String,
+    targetVariables: Option[Seq[Expression]]): LogicalPlan = {
+    // If targetVariables is defined, statement needs to be a query.
```

Comment:
non-query command can also return value, e.g. `SHOW TABLE`, and DML commands can also return value like number of rows added/deleted. Why do we have this limitation? cc @srielau as well

Reply:
I think what matters is the logical plan schema which needs to match the target variables, and the plan can be anything.

Reply:
I'll let @srielau  comment

Reply:
we can also leave it for follow up PR